src/service_discovery_setup.js

```diff
@@ -0,0 +1,274 @@
+  this.serviceState = !!state;    // Make sure it's a boolean value
+  if (state) {
+    this.dnssdHandle.start();
+  } else {
+    this.dnssdHandle.stop();
+  }
+  console.log(`Service Discover state changed to: ${state}`);
+};
+
+
+/**
+ * A small helper method.
+ *
+ * Used to update the local name of the mDNS service discovery. It does this
+ * by updating it's localDomain name, stopping the old service and then
+ * creating a new dnssd object.
+ *
+ * @param {local DNS name. e.g. Myhome-iot}
+ */
+
+DNSserviceDiscovery.prototype.setLocalDomain = function(localDomain) {
+  // TODO check the localDomain name is a legit mDNS name
+  // Check any letter or number a-Z,A-Z, 0-9 and '-' any number of times with
+  // a length less than 63 or contain www and api
+  localDomain = localDomain.toLowerCase();
+  const re = new RegExp(/^([a-z0-9]|[a-z0-9][a-z0-9-]*[a-z0-9])$/);
+  const valid = re.test(localDomain) && !['api', 'www'].includes(localDomain) &&
+    localDomain.length <= 63;
+
+  if (valid) {
+    // Change our object data member, and stop the old service first.
+    console.log(`Service Discover local Domain validity is: ${valid}`);
+    console.log(`Service Discover localDomain changed to: ${localDomain} `);
+    this.localDomain = localDomain;
+    this.dnssdHandle.stop();
+
+    // Now let's start a new service with the new name.
+    const txt = {
+      desc: this.description,
+      protocols: this.protocol,
+      power: this.power,
+    };
+    const options = {
+      name: this.localName,
+      host: this.localDomain,
+      txt: txt
+    };
+
+    this.dnssdHandle = dnssd.Advertisement(dnssd.tcp('http'),
+                                           this.port, options);
+  } else {
+    console.error(`Our Domain Name string did not match [a-z,A-Z,0-9,-] 
+    or use www || api. We got: ${localDomain}`);
+    throw new Error(`Invalid local domain name. It should only consist 
+    of characters a-Z, A-Z, 0-9, - or use www, api`);
+  }
+};
+
+
+/**
+ * changeProfile method.
+ *
+ * This method will take a new profile 'object' and use this to update
+ * the objects internal variables. It then will create a new mDNS
+ * advertiser object using the new profile and restart the advertiser
+ * depending on the state of the DNSserviceDiscovery manager.
+ *
+ * Returns:
+ *
+ * @param {object} newProfile
+ *          e.g newProfile = { name: "Mozilla IoT Gateway",
+ *                              host: "mozillaGateway",     // Local DNS name
+ *                              txt: {  desc: "descriptive text",
+ *                                      protocols:  "http, https, oauth2 etc..",
+ *                                      power: "default power"
+ *                                    }
+ *                            }
+ */
+DNSserviceDiscovery.prototype.changeProfile = function(newProfile) {
+  try {
+    console.log(`Service Discovery is changing profile. The localDomain is 
```

Comment:
Same comment as above about line breaks inside template strings.

Reply:
i have used the '\' method. but stacktrace is not telling me what is the correct method... i suggest nodejs people do this:

/> python
/> >>> import this
/> >>>There should be one-- and preferably only one --obvious way to do it.

:-)